File: untitled folder/untitled folder 2/flutter_secure_biometrics/android/src/main/kotlin/com/example/flutter_secure_biometrics/BehavioralAnalyzer.kt
package com.example.flutter_secure_biometrics

import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import android.os.Handler
import android.os.Looper
import kotlinx.coroutines.*
import java.util.concurrent.ConcurrentHashMap
import kotlin.math.sqrt

class BehavioralAnalyzer : SensorEventListener {
    
    private var sensorManager: SensorManager? = null
    private var accelerometer: Sensor? = null
    private var gyroscope: Sensor? = null
    private var isMonitoring = false
    
    private val behavioralData = ConcurrentHashMap<String, MutableList<Double>>()
    private val trustScoreHistory = mutableListOf<Double>()
    private var currentTrustScore = 1.0
    private var currentAppId: String? = null
    
    private val monitoringScope = CoroutineScope(Dispatchers.Default + SupervisorJob())
    private val handler = Handler(Looper.getMainLooper())
    
    // Behavioral pattern baselines
    private val typingSpeedBaseline = mutableListOf<Double>()
    private val touchPressureBaseline = mutableListOf<Double>()
    private val deviceMovementBaseline = mutableListOf<Double>()
    
    fun startContinuousMonitoring(context: Context, config: Map<String, Any>, appId: String): Boolean {
        return try {
            currentAppId = appId
            sensorManager = context.getSystemService(Context.SENSOR_SERVICE) as SensorManager
            accelerometer = sensorManager?.getDefaultSensor(Sensor.TYPE_ACCELEROMETER)
            gyroscope = sensorManager?.getDefaultSensor(Sensor.TYPE_GYROSCOPE)
            
            // Register sensor listeners
            accelerometer?.let {
                sensorManager?.registerListener(this, it, SensorManager.SENSOR_DELAY_NORMAL)
            }
            gyroscope?.let {
                sensorManager?.registerListener(this, it, SensorManager.SENSOR_DELAY_NORMAL)
            }
            
            isMonitoring = true
            startTrustScoreCalculation()
            
            true
        } catch (e: Exception) {
            false
        }
    }
    
    fun stopContinuousMonitoring() {
        isMonitoring = false
        sensorManager?.unregisterListener(this)
        monitoringScope.cancel()
        currentAppId = null
    }
    
    fun getCurrentTrustScore(): Map<String, Any> {
        return mapOf(
            "value" to currentTrustScore,
            "timestamp" to System.currentTimeMillis(),
            "factors" to mapOf(
                "deviceMovement" to getDeviceMovementScore(),
                "behavioralConsistency" to getBehavioralConsistencyScore(),
                "temporalPattern" to getTemporalPatternScore()
            ),
            "reason" to getTrustScoreReason()
        )
    }
    
    fun updateMetrics(metrics: Map<String, Any>): Boolean {
        return try {
            // Update typing pattern metrics
            (metrics["typingPattern"] as? Map<String, Any>)?.let { typing ->
                (typing["averageSpeed"] as? Double)?.let { speed ->
                    updateTypingSpeed(speed)
                }
                (typing["pressure"] as? Double)?.let { pressure ->
                    updateTouchPressure(pressure)
                }
            }
            
            // Update touch gesture metrics
            (metrics["touchGesture"] as? Map<String, Any>)?.let { touch ->
                (touch["pressureSensitivity"] as? Double)?.let { pressure ->
                    updateTouchPressure(pressure)
                }
            }
            
            // Recalculate trust score
            calculateTrustScore()
            
            true
        } catch (e: Exception) {
            false
        }
    }
    
    fun lockApplication() {
        currentTrustScore = 0.0
        isMonitoring = false
    }
    
    fun exportTrainingData(): Map<String, Any> {
        return mapOf(
            "typingSpeedBaseline" to typingSpeedBaseline.toList(),
            "touchPressureBaseline" to touchPressureBaseline.toList(),
            "deviceMovementBaseline" to deviceMovementBaseline.toList(),
            "trustScoreHistory" to trustScoreHistory.toList(),
            "appId" to (currentAppId ?: ""),
            "exportTimestamp" to System.currentTimeMillis()
        )
    }
    
    fun importTrainingData(data: Map<String, Any>): Boolean {
        return try {
            (data["typingSpeedBaseline"] as? List<Double>)?.let {
                typingSpeedBaseline.clear()
                typingSpeedBaseline.addAll(it)
            }
            
            (data["touchPressureBaseline"] as? List<Double>)?.let {
                touchPressureBaseline.clear()
                touchPressureBaseline.addAll(it)
            }
            
            (data["deviceMovementBaseline"] as? List<Double>)?.let {
                deviceMovementBaseline.clear()
                deviceMovementBaseline.addAll(it)
            }
            
            true
        } catch (e: Exception) {
            false
        }
    }
    
    override fun onSensorChanged(event: SensorEvent?) {
        if (!isMonitoring || event == null) return
        
        when (event.sensor.type) {
            Sensor.TYPE_ACCELEROMETER -> {
                val magnitude = sqrt(
                    event.values[0] * event.values[0] +
                    event.values[1] * event.values[1] +
                    event.values[2] * event.values[2]
                )
                updateDeviceMovement(magnitude.toDouble())
            }
            
            Sensor.TYPE_GYROSCOPE -> {
                val rotationMagnitude = sqrt(
                    event.values[0] * event.values[0] +
                    event.values[1] * event.values[1] +
                    event.values[2] * event.values[2]
                )
                updateDeviceRotation(rotationMagnitude.toDouble())
            }
        }
    }
    
    override fun onAccuracyChanged(sensor: Sensor?, accuracy: Int) {
        // Handle sensor accuracy changes if needed
    }
    
    private fun startTrustScoreCalculation() {
        monitoringScope.launch {
            while (isMonitoring) {
                calculateTrustScore()
                delay(5000) // Update every 5 seconds
            }
        }
    }
    
    private fun calculateTrustScore() {
        val factors = mutableMapOf<String, Double>()
        
        // Device movement consistency
        factors["deviceMovement"] = getDeviceMovementScore()
        
        // Behavioral consistency
        factors["behavioralConsistency"] = getBehavioralConsistencyScore()
        
        // Temporal patterns
        factors["temporalPattern"] = getTemporalPatternScore()
        
        // Calculate weighted average
        val weights = mapOf(
            "deviceMovement" to 0.3,
            "behavioralConsistency" to 0.5,
            "temporalPattern" to 0.2
        )
        
        currentTrustScore = factors.entries.sumOf { (key, value) ->
            value * (weights[key] ?: 0.0)
        }
        
        // Ensure trust score is between 0 and 1
        currentTrustScore = currentTrustScore.coerceIn(0.0, 1.0)
        
        // Add to history
        trustScoreHistory.add(currentTrustScore)
        if (trustScoreHistory.size > 100) {
            trustScoreHistory.removeAt(0)
        }
    }
    
    private fun getDeviceMovementScore(): Double {
        val movementData = behavioralData["deviceMovement"] ?: return 1.0
        if (movementData.isEmpty() || deviceMovementBaseline.isEmpty()) return 1.0
        
        val recentMovement = movementData.takeLast(10).average()
        val baselineMovement = deviceMovementBaseline.average()
        
        // Calculate similarity score (closer to baseline = higher score)
        val difference = kotlin.math.abs(recentMovement - baselineMovement)
        val maxDifference = baselineMovement * 0.5 // Allow 50% variance
        
        return (1.0 - (difference / maxDifference)).coerceIn(0.0, 1.0)
    }
    
    private fun getBehavioralConsistencyScore(): Double {
        if (typingSpeedBaseline.isEmpty() && touchPressureBaseline.isEmpty()) return 1.0
        
        var consistencyScore = 0.0
        var factorCount = 0
        
        // Typing speed consistency
        if (typingSpeedBaseline.isNotEmpty()) {
            val typingData = behavioralData["typingSpeed"] ?: emptyList()
            if (typingData.isNotEmpty()) {
                val recentTyping = typingData.takeLast(5).average()
                val baselineTyping = typingSpeedBaseline.average()
                val typingConsistency = 1.0 - (kotlin.math.abs(recentTyping - baselineTyping) / baselineTyping).coerceAtMost(1.0)
                consistencyScore += typingConsistency
                factorCount++
            }
        }
        
        // Touch pressure consistency
        if (touchPressureBaseline.isNotEmpty()) {
            val pressureData = behavioralData["touchPressure"] ?: emptyList()
            if (pressureData.isNotEmpty()) {
                val recentPressure = pressureData.takeLast(5).average()
                val baselinePressure = touchPressureBaseline.average()
                val pressureConsistency = 1.0 - (kotlin.math.abs(recentPressure - baselinePressure) / baselinePressure).coerceAtMost(1.0)
                consistencyScore += pressureConsistency
                factorCount++
            }
        }
        
        return if (factorCount > 0) consistencyScore / factorCount else 1.0
    }
    
    private fun getTemporalPatternScore(): Double {
        // Analyze temporal patterns in user behavior
        if (trustScoreHistory.size < 10) return 1.0
        
        val recentScores = trustScoreHistory.takeLast(10)
        val variance = calculateVariance(recentScores)
        
        // Lower variance indicates more consistent behavior
        return (1.0 - variance).coerceIn(0.0, 1.0)
    }
    
    private fun getTrustScoreReason(): String {
        return when {
            currentTrustScore >= 0.9 -> "Excellent behavioral match"
            currentTrustScore >= 0.8 -> "Good behavioral consistency"
            currentTrustScore >= 0.7 -> "Acceptable behavioral patterns"
            currentTrustScore >= 0.6 -> "Some behavioral inconsistencies detected"
            else -> "Significant behavioral anomalies detected"
        }
    }
    
    private fun updateTypingSpeed(speed: Double) {
        val typingData = behavioralData.getOrPut("typingSpeed") { mutableListOf() }
        typingData.add(speed)
        if (typingData.size > 50) typingData.removeAt(0)
        
        // Update baseline if we have enough data
        if (typingSpeedBaseline.size < 20) {
            typingSpeedBaseline.add(speed)
        }
    }
    
    private fun updateTouchPressure(pressure: Double) {
        val pressureData = behavioralData.getOrPut("touchPressure") { mutableListOf() }
        pressureData.add(pressure)
        if (pressureData.size > 50) pressureData.removeAt(0)
        
        // Update baseline if we have enough data
        if (touchPressureBaseline.size < 20) {
            touchPressureBaseline.add(pressure)
        }
    }
    
    private fun updateDeviceMovement(movement: Double) {
        val movementData = behavioralData.getOrPut("deviceMovement") { mutableListOf() }
        movementData.add(movement)
        if (movementData.size > 100) movementData.removeAt(0)
        
        // Update baseline if we have enough data
        if (deviceMovementBaseline.size < 50) {
            deviceMovementBaseline.add(movement)
        }
    }
    
    private fun updateDeviceRotation(rotation: Double) {
        val rotationData = behavioralData.getOrPut("deviceRotation") { mutableListOf() }
        rotationData.add(rotation)
        if (rotationData.size > 100) rotationData.removeAt(0)
    }
    
    private fun calculateVariance(values: List<Double>): Double {
        if (values.isEmpty()) return 0.0
        val mean = values.average()
        val squaredDifferences = values.map { (it - mean) * (it - mean) }
        return squaredDifferences.average()
    }
}

File: untitled folder/untitled folder 2/flutter_secure_biometrics/ios/Classes/BehavioralAnalyzer.swift
import Foundation
import CoreMotion
import Flutter
#if os(iOS)
import UIKit
#endif

class BehavioralAnalyzer {
    
    private let motionManager = CMMotionManager()
    private var isMonitoring = false
    private var currentAppId: String?
    private var eventSink: FlutterEventSink?
    
    private var behavioralData: [String: [Double]] = [:]
    private var trustScoreHistory: [Double] = []
    private var currentTrustScore: Double = 1.0
    
    // Behavioral pattern baselines
    private var typingSpeedBaseline: [Double] = []
    private var touchPressureBaseline: [Double] = []
    private var deviceMovementBaseline: [Double] = []
    
    private var monitoringTimer: Timer?
    
    func startContinuousMonitoring(config: [String: Any], appId: String) -> Bool {
        guard !isMonitoring else { return false }
        
        currentAppId = appId
        isMonitoring = true
        
        // Start motion sensors
        startMotionMonitoring()
        
        // Start trust score calculation timer
        startTrustScoreCalculation()
        
        return true
    }
    
    func stopContinuousMonitoring() {
        isMonitoring = false
        currentAppId = nil
        
        motionManager.stopAccelerometerUpdates()
        motionManager.stopGyroUpdates()
        motionManager.stopDeviceMotionUpdates()
        
        monitoringTimer?.invalidate()
        monitoringTimer = nil
    }
    
    func getCurrentTrustScore() -> [String: Any] {
        return [
            "value": currentTrustScore,
            "timestamp": Date().timeIntervalSince1970 * 1000,
            "factors": [
                "deviceMovement": getDeviceMovementScore(),
                "behavioralConsistency": getBehavioralConsistencyScore(),
                "temporalPattern": getTemporalPatternScore()
            ],
            "reason": getTrustScoreReason()
        ]
    }
    
    func updateMetrics(metrics: [String: Any]) -> Bool {
        // Update typing pattern metrics
        if let typingPattern = metrics["typingPattern"] as? [String: Any] {
            if let speed = typingPattern["averageSpeed"] as? Double {
                updateTypingSpeed(speed)
            }
            if let pressure = typingPattern["pressure"] as? Double {
                updateTouchPressure(pressure)
            }
        }
        
        // Update touch gesture metrics
        if let touchGesture = metrics["touchGesture"] as? [String: Any] {
            if let pressure = touchGesture["pressureSensitivity"] as? Double {
                updateTouchPressure(pressure)
            }
        }
        
        // Recalculate trust score
        calculateTrustScore()
        
        return true
    }
    
    func lockApplication() {
        currentTrustScore = 0.0
        isMonitoring = false
        
        // Send security alert
        eventSink?([
            "type": "securityAlert",
            "data": [
                "alertType": "applicationLocked",
                "reason": "Trust score too low",
                "timestamp": Date().timeIntervalSince1970 * 1000
            ]
        ])
    }
    
    func exportTrainingData() -> [String: Any] {
        return [
            "typingSpeedBaseline": typingSpeedBaseline,
            "touchPressureBaseline": touchPressureBaseline,
            "deviceMovementBaseline": deviceMovementBaseline,
            "trustScoreHistory": trustScoreHistory,
            "appId": currentAppId ?? "",
            "exportTimestamp": Date().timeIntervalSince1970 * 1000
        ]
    }
    
    func importTrainingData(data: [String: Any]) -> Bool {
        if let typingBaseline = data["typingSpeedBaseline"] as? [Double] {
            typingSpeedBaseline = typingBaseline
        }
        
        if let touchBaseline = data["touchPressureBaseline"] as? [Double] {
            touchPressureBaseline = touchBaseline
        }
        
        if let movementBaseline = data["deviceMovementBaseline"] as? [Double] {
            deviceMovementBaseline = movementBaseline
        }
        
        return true
    }
    
    func setEventSink(eventSink: FlutterEventSink?) {
        self.eventSink = eventSink
    }
    
    // MARK: - Private Methods
    
    private func startMotionMonitoring() {
        guard motionManager.isAccelerometerAvailable && motionManager.isGyroAvailable else { return }
        
        motionManager.accelerometerUpdateInterval = 0.1
        motionManager.gyroUpdateInterval = 0.1
        
        // Start accelerometer updates
        motionManager.startAccelerometerUpdates(to: .main) { [weak self] data, error in
            guard let self = self, let data = data, self.isMonitoring else { return }
            
            let magnitude = sqrt(data.acceleration.x * data.acceleration.x +
                               data.acceleration.y * data.acceleration.y +
                               data.acceleration.z * data.acceleration.z)
            
            self.updateDeviceMovement(magnitude)
        }
        
        // Start gyroscope updates
        motionManager.startGyroUpdates(to: .main) { [weak self] data, error in
            guard let self = self, let data = data, self.isMonitoring else { return }
            
            let rotationMagnitude = sqrt(data.rotationRate.x * data.rotationRate.x +
                                       data.rotationRate.y * data.rotationRate.y +
                                       data.rotationRate.z * data.rotationRate.z)
            
            self.updateDeviceRotation(rotationMagnitude)
        }
    }
    
    private func startTrustScoreCalculation() {
        monitoringTimer = Timer.scheduledTimer(withTimeInterval: 5.0, repeats: true) { [weak self] _ in
            guard let self = self, self.isMonitoring else { return }
            
            self.calculateTrustScore()
            
            // Send trust score update
            let trustScore = self.getCurrentTrustScore()
            self.eventSink?([
                "type": "trustScore",
                "data": trustScore
            ])
        }
    }
    
    private func calculateTrustScore() {
        var factors: [String: Double] = [:]
        
        // Device movement consistency
        factors["deviceMovement"] = getDeviceMovementScore()
        
        // Behavioral consistency
        factors["behavioralConsistency"] = getBehavioralConsistencyScore()
        
        // Temporal patterns
        factors["temporalPattern"] = getTemporalPatternScore()
        
        // Calculate weighted average
        let weights: [String: Double] = [
            "deviceMovement": 0.3,
            "behavioralConsistency": 0.5,
            "temporalPattern": 0.2
        ]
        
        currentTrustScore = factors.reduce(0.0) { result, factor in
            result + (factor.value * (weights[factor.key] ?? 0.0))
        }
        
        // Ensure trust score is between 0 and 1
        currentTrustScore = max(0.0, min(1.0, currentTrustScore))
        
        // Add to history
        trustScoreHistory.append(currentTrustScore)
        if trustScoreHistory.count > 100 {
            trustScoreHistory.removeFirst()
        }
    }
    
    private func getDeviceMovementScore() -> Double {
        guard let movementData = behavioralData["deviceMovement"],
              !movementData.isEmpty,
              !deviceMovementBaseline.isEmpty else { return 1.0 }
        
        let recentMovement = Array(movementData.suffix(10)).reduce(0, +) / Double(min(10, movementData.count))
        let baselineMovement = deviceMovementBaseline.reduce(0, +) / Double(deviceMovementBaseline.count)
        
        // Calculate similarity score (closer to baseline = higher score)
        let difference = abs(recentMovement - baselineMovement)
        let maxDifference = baselineMovement * 0.5 // Allow 50% variance
        
        return max(0.0, min(1.0, 1.0 - (difference / maxDifference)))
    }
    
    private func getBehavioralConsistencyScore() -> Double {
        guard !typingSpeedBaseline.isEmpty || !touchPressureBaseline.isEmpty else { return 1.0 }
        
        var consistencyScore = 0.0
        var factorCount = 0
        
        // Typing speed consistency
        if !typingSpeedBaseline.isEmpty,
           let typingData = behavioralData["typingSpeed"],
           !typingData.isEmpty {
            let recentTyping = Array(typingData.suffix(5)).reduce(0, +) / Double(min(5, typingData.count))
            let baselineTyping = typingSpeedBaseline.reduce(0, +) / Double(typingSpeedBaseline.count)
            let typingConsistency = max(0.0, 1.0 - abs(recentTyping - baselineTyping) / baselineTyping)
            consistencyScore += typingConsistency
            factorCount += 1
        }
        
        // Touch pressure consistency
        if !touchPressureBaseline.isEmpty,
           let pressureData = behavioralData["touchPressure"],
           !pressureData.isEmpty {
            let recentPressure = Array(pressureData.suffix(5)).reduce(0, +) / Double(min(5, pressureData.count))
            let baselinePressure = touchPressureBaseline.reduce(0, +) / Double(touchPressureBaseline.count)
            let pressureConsistency = max(0.0, 1.0 - abs(recentPressure - baselinePressure) / baselinePressure)
            consistencyScore += pressureConsistency
            factorCount += 1
        }
        
        return factorCount > 0 ? consistencyScore / Double(factorCount) : 1.0
    }
    
    private func getTemporalPatternScore() -> Double {
        guard trustScoreHistory.count >= 10 else { return 1.0 }
        
        let recentScores = Array(trustScoreHistory.suffix(10))
        let variance = calculateVariance(recentScores)
        
        // Lower variance indicates more consistent behavior
        return max(0.0, min(1.0, 1.0 - variance))
    }
    
    private func getTrustScoreReason() -> String {
        switch currentTrustScore {
        case 0.9...1.0:
            return "Excellent behavioral match"
        case 0.8..<0.9:
            return "Good behavioral consistency"
        case 0.7..<0.8:
            return "Acceptable behavioral patterns"
        case 0.6..<0.7:
            return "Some behavioral inconsistencies detected"
        default:
            return "Significant behavioral anomalies detected"
        }
    }
    
    private func updateTypingSpeed(_ speed: Double) {
        if behavioralData["typingSpeed"] == nil {
            behavioralData["typingSpeed"] = []
        }
        behavioralData["typingSpeed"]?.append(speed)
        
        if behavioralData["typingSpeed"]!.count > 50 {
            behavioralData["typingSpeed"]?.removeFirst()
        }
        
        // Update baseline if we have enough data
        if typingSpeedBaseline.count < 20 {
            typingSpeedBaseline.append(speed)
        }
    }
    
    private func updateTouchPressure(_ pressure: Double) {
        if behavioralData["touchPressure"] == nil {
            behavioralData["touchPressure"] = []
        }
        behavioralData["touchPressure"]?.append(pressure)
        
        if behavioralData["touchPressure"]!.count > 50 {
            behavioralData["touchPressure"]?.removeFirst()
        }
        
        // Update baseline if we have enough data
        if touchPressureBaseline.count < 20 {
            touchPressureBaseline.append(pressure)
        }
    }
    
    private func updateDeviceMovement(_ movement: Double) {
        if behavioralData["deviceMovement"] == nil {
            behavioralData["deviceMovement"] = []
        }
        behavioralData["deviceMovement"]?.append(movement)
        
        if behavioralData["deviceMovement"]!.count > 100 {
            behavioralData["deviceMovement"]?.removeFirst()
        }
        
        // Update baseline if we have enough data
        if deviceMovementBaseline.count < 50 {
            deviceMovementBaseline.append(movement)
        }
    }
    
    private func updateDeviceRotation(_ rotation: Double) {
        if behavioralData["deviceRotation"] == nil {
            behavioralData["deviceRotation"] = []
        }
        behavioralData["deviceRotation"]?.append(rotation)
        
        if behavioralData["deviceRotation"]!.count > 100 {
            behavioralData["deviceRotation"]?.removeFirst()
        }
    }
    
    private func calculateVariance(_ values: [Double]) -> Double {
        guard !values.isEmpty else { return 0.0 }
        
        let mean = values.reduce(0, +) / Double(values.count)
        let squaredDifferences = values.map { pow($0 - mean, 2) }
        return squaredDifferences.reduce(0, +) / Double(values.count)
    }
}
